messaging/messaging/src/main/java/io/helidon/messaging/MessagingImpl.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c)  2020 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.messaging;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Flow;
+
+import io.helidon.common.configurable.ThreadPoolSupplier;
+import io.helidon.common.reactive.Multi;
+import io.helidon.config.Config;
+import io.helidon.config.ConfigSources;
+import io.helidon.config.ConfigValue;
+
+import org.eclipse.microprofile.reactive.messaging.spi.Connector;
+import org.eclipse.microprofile.reactive.messaging.spi.ConnectorFactory;
+import org.eclipse.microprofile.reactive.messaging.spi.IncomingConnectorFactory;
+import org.eclipse.microprofile.reactive.messaging.spi.OutgoingConnectorFactory;
+
+class MessagingImpl implements Messaging {
+
+    private final Set<Emitter<?>> emitters = new HashSet<>();
+    private final Map<String, Channel<?>> channelMap = new HashMap<>();
+    private final Map<String, IncomingConnectorFactory> incomingConnectors = new HashMap<>();
+    private final Map<String, OutgoingConnectorFactory> outgoingConnectors = new HashMap<>();
+    private Config config;
+    private MessagingImpl.State state = MessagingImpl.State.INIT;
+    private ThreadPoolSupplier threadPoolSupplier;
+
+    MessagingImpl() {
+    }
+
+    public void start() {
+        state.start(this);
+        threadPoolSupplier = ThreadPoolSupplier.builder()
```

Comment:
Create a thread pool only if at least one emitter is configured.
Get it before the `forEach` pool and store the executor service instance.
Close it when closing meessaging.

Reply:
use an identifier of messaging for
 - thread name prefix
 - helidon context

Reply:
Make sure there is:

1. There is a "messaging" `io.helidon.common.Context` instance for the messaging instance
2. There is a "message" `Context` for each message flowing through the messaging (the same one from message producer to message consumer(s), with parent configured to 1.

Reply:
fixed, context is unique for every message see `io.helidon.messaging.ContextSubscriber`